File: src/main/java/com/coalesce/utils/TemperatureUnit.java
package com.coalesce.utils;

public enum TemperatureUnit {
    FAHRENHEIT() {
        @Override
        public double toCelsius(double from) {
            return (from - 32) / 1.8;
        }

        @Override
        public double toFahrenheit(double from) {
            return from;
        }

        @Override
        public double toKelvin(double from) {
            return (from + 459.69) / 1.8;
        }

        @Override
        public double convert(double from, TemperatureUnit unit) {
            return unit.toFahrenheit(from);
        }
    },
    CELSIUS() {
        public double toCelsius(double from) {
            return from;
        }

        public double toFahrenheit(double from) {
            return (from * 1.8) + 32;
        }

        public double toKelvin(double from) {
            return from + 273.15;
        }

        public double convert(double from, TemperatureUnit unit) {
            return unit.toCelsius(from);
        }
    },
    KELVIN() {
        @Override
        public double toCelsius(double from) {
            return from - 273.15;
        }

        @Override
        public double toKelvin(double from) {
            return from;
        }

        @Override
        public double toFahrenheit(double from) {
            return (from * 1.8) - 459.67;
        }

        @Override
        public double convert(double from, TemperatureUnit unit) {
            return unit.toKelvin(from);
        }
    };

    public double toCelsius(double from) {
        throw new AbstractMethodError();
    }

    public double toFahrenheit(double from) {
        throw new AbstractMethodError();
    }

    public double toKelvin(double from) {
        throw new AbstractMethodError();
    }

    public double convert(double from, TemperatureUnit unit) {
        throw new AbstractMethodError();
    }
}

File: src/main/kotlin/com/coalesce/utils/TemperatureUnit.kt
package com.coalesce.utils

enum class TemperatureUnit {
    KELVIN() {
        override fun toCelsius(from: Double): Double {
            return from - 273.15
        }

        override fun toKelvin(from: Double): Double {
            return from
        }

        override fun toFahrenheit(from: Double): Double {
            return from * 1.8 - 459.67
        }

        override fun convert(from: Double, unit: TemperatureUnit): Double {
            return unit.toKelvin(from)
        }
    },
    CELSIUS() {
        override fun toCelsius(from: Double): Double {
            return from
        }

        override fun toFahrenheit(from: Double): Double {
            return from * 1.8 + 32
        }

        override fun toKelvin(from: Double): Double {
            return from + 273.15
        }

        override fun convert(from: Double, unit: TemperatureUnit): Double {
            return unit.toCelsius(from)
        }
    },
    FAHRENHEIT() {
        override fun toCelsius(from: Double): Double {
            return (from - 32) / 1.8
        }

        override fun toFahrenheit(from: Double): Double {
            return from
        }

        override fun toKelvin(from: Double): Double {
            return (from + 459.69) / 1.8
        }

        override fun convert(from: Double, unit: TemperatureUnit): Double {
            return unit.toFahrenheit(from)
        }
    };

    open fun toCelsius(from: Double): Double {
        throw AbstractMethodError()
    }

    open fun toFahrenheit(from: Double): Double {
        throw AbstractMethodError()
    }

    open fun toKelvin(from: Double): Double {
        throw AbstractMethodError()
    }

    open fun convert(from: Double, unit: TemperatureUnit): Double {
        throw AbstractMethodError()
    }
}